src/dos/dos_keyboard_layout.cpp

```diff
@@ -39,78 +39,99 @@
 
 extern void DOS_UpdateCurrentProgramName();
 
-static void notify_code_page_changed(const bool keyboard_layout_changed = false)
+static const std::string ResourceDirKeyboard = "freedos-keyboard";
+static const std::string ResourceDirCodePage = "freedos-cpi";
+
+static void notify_code_page_changed()
 {
 	// Re-create various information to match new code page
 	DOS_UpdateCurrentProgramName();
-	DOS_RefreshCountryInfo(keyboard_layout_changed);
+	DOS_RepopulateCountryInfo();
 	AUTOEXEC_NotifyNewCodePage();
 }
 
 // A common pattern in the keyboard layout file is to try opening the requested
-// file first within DOS, then from the local path, and finally from builtin
-// resources. This function performs those in order and returns the first hit.
-static FILE_unique_ptr open_layout_file(const char *name, const char *resource_dir = nullptr)
+// file first within DOS, then from the local path.
+// This function performs those in order and returns the first hit.
+static FILE_unique_ptr open_layout_file(const std::string& name)
 {
-	constexpr auto file_perms = "rb";
+	constexpr auto FilePermissions = "rb";
 
 	// Try opening from DOS first (this can throw, so we catch/ignore)
 	uint8_t drive = 0;
 	char fullname[DOS_PATHLENGTH] = {};
-	if (DOS_MakeName(name, fullname, &drive)) try {
-		// try to open file on mounted drive first
-		const auto ldp = std::dynamic_pointer_cast<localDrive>(Drives[drive]);
-		if (ldp) {
-			if (const auto fp = ldp->GetHostFilePtr(fullname, file_perms); fp) {
-				return FILE_unique_ptr(fp);
+	if (DOS_MakeName(name.c_str(), fullname, &drive)) {
+		try {
+			// try to open file on mounted drive first
+			const auto ldp = std::dynamic_pointer_cast<localDrive>(
+			        Drives[drive]);
+			if (ldp) {
+				if (const auto fp = ldp->GetHostFilePtr(fullname,
+				                                        FilePermissions);
+				    fp) {
+					return FILE_unique_ptr(fp);
+				}
 			}
+		} catch (...) {
 		}
-	} catch (...) {}
+	}
 
 	// Then try from the local filesystem
-	if (const auto fp = fopen(name, file_perms); fp)
+	if (const auto fp = fopen(name.c_str(), FilePermissions); fp) {
 		return FILE_unique_ptr(fp);
-
-	// Finally try from a built-in resources
-	if (resource_dir)
-		if (const auto rp = get_resource_path(resource_dir, name); !rp.empty())
-			return make_fopen(rp.string().c_str(), file_perms);
+	}
 
 	return nullptr;
 }
 
+// Try to open file from bundled resources
+static FILE_unique_ptr open_bundled_file(const std::string& name,
```

Comment:
Yeah okay, so maybe you'd need `BundledFile` as well, not just `Layout`.

Reply:
Refactoring of the keyboard layout and CPI file loading is planned, but out of scope of this PR - it is already huge. 

See tickets https://github.com/dosbox-staging/dosbox-staging/issues/1925 and https://github.com/dosbox-staging/dosbox-staging/issues/4004.